C19016 accord table config

bdeggleston opened the pull request.

maedhroz committed:
CEP-15: Multi-Partition Transaction CQL Support (Alpha)

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-17719

Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>
Co-authored-by: Blake Eggleston <bdeggleston@gmail.com>
Co-authored-by: David Capwell <dcapwell@apache.org>

belliottsmith committed:
Partial Replication of Transaction State

The replication of transactions and their processing state must be optimal, i.e. linearly proportional to the size of the transaction and the number of keys involved. This patch ensures a transaction is replicated in slices, with each slice being written (replication factor) times only to the shard that owns the keys, and only once (except in the case the transaction is recovered).

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-17109

bdeggleston committed:
CEP-15 (C*): Messaging and storage engine integration

patch by Blake Eggleston; reviewed by Benedict Elliott Smith, David Capwell for CASSANDRA-17103

belliottsmith committed:
CEP-15: Routables
  - Integrate accord-core changes for CASSANDRA-18087

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18087

dcapwell committed:
Command table now uses a local version added to the payload rather than rely on a version column, and fixed a few feedback related issues

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-17103

belliottsmith committed:
Transaction Expiration

Ensure old or abandoned transactions are invalidated wherever possible.

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18041

belliottsmith committed:
Invalidation fixes/improvements
 - Integrate accord-core changes for CASSANDRA-18057

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18057

belliottsmith committed:
Refactor response tracking to improve efficiency and clarity; introduce dedicated property tests; re-activate fast-path during range movements

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18056

aweisberg committed:
Support CAS and serial read on Accord

patch by Ariel Weisberg; Reviewed by Caleb Rackliffe for CASSANDRA-18100

belliottsmith committed:
Shard local CommandStores on contiguous ranges

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18142

belliottsmith committed:
Refactor Timestamp/TxnId
 - Combine real and logical into a single 64-but HLC
 - Introduce 16 flag bits
 - Pack epoch (48-bits), HLC (64-bits) and flags (16-bits) into two longs in memory

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18172

belliottsmith committed:
Introduce RangeDeps

Refactor Deps into KeyDeps and RangeDeps

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18173

belliottsmith committed:
Introduce Range transactions

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18174

belliottsmith committed:
ninja: pick a cassandra-accord SHA to build against to decouple source trees

dcapwell committed:
CASSANDRA-18154: CEP-15: Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

dcapwell committed:
CEP-15: (C*) Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18154

aweisberg committed:
Accord support for live migration

Patch by Ariel Weisberg; Reviewed by Blake Eggleston for CASSANDRA-18129

maedhroz committed:
Post-rebase fixes around CASSANDRA-18099 and CASSANDRA-18114 (among other things)

bdeggleston committed:
add AsyncChain implementations and tests

Patch by Blake Eggleston; Reviewed by David Capwell & Benedict Elliott Smith for CASSANDRA-18004

maedhroz committed:
Feature Flag for Accord Transactions

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-18195

iamaleksey committed:
Switch Node.Id from long to int

patch by Aleksey Yeschenko; reviewed by Benedict Elliott Smith for
CASSANDRA-18135

dcapwell committed:
Ninja: CASSANDRA-18214 added rat checks to Accord, but our metadata files .BRANCH and .REPO do not have a license, causing the build to fail; exclude rat when building accord

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18214

maedhroz committed:
use equals() rather than reference equality for excluding ourselves from the dependency builder

patch by Benedict; reviewed by Caleb Rackliffe for CASSANDRA-18218

bdeggleston committed:
CEP-15/Accord Immutable State Refactor

Patch by Blake Eggleston; Reviewed by David Capwell and Benedict Elliott Smith for Cassandra-18192

jacek-lewandowski committed:
Fix null value handling for static columns

patch by <jacek-lewandowski>; reviewed by <maedhroz> and <dcapwell> for CASSANDRA-18241

iamaleksey committed:
CEP-15: Store PreAccept, Accept, Commit, and Apply messages
in a durable log before processing by CommandStores

patch by Aleksey Yeschenko; reviewed by David Capwell for
CASSANDRA-18344

jacek-lewandowski committed:
Fix statement validation against partition range queries

patch by Jacek Lewandowski; reviewed by Caleb Rackliffe for CASSANDRA-18240

dcapwell committed:
CEP-15: (C*) Add git submodule for Accord

patch by David Capwell; reviewed by Caleb Rackliffe, Michael Semb Wever for CASSANDRA-18204

dcapwell committed:
Switched away from https endpoints for git submodules in favor of relative paths during development

patch by David Capwell; reviewed by Caleb Rackliffe, Michael Semb Wever for CASSANDRA-18204

dcapwell committed:
Add support for prepared statements for accord transactions

patch by David Capwell; reviewed by Alex Petrov, Caleb Rackliffe, Jacek Lewandowski for CASSANDRA-18299

dcapwell committed:
Remove git hook for pre-push as it is redundant and causes issues when merging to mainline

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18309

dcapwell committed:
CEP-15: (Accord) Migrate Accord away from JDK random to a new interface RandomSource

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-18213

jacek-lewandowski committed:
Improve transaction statement validation

patch by Jacek Lewandowski; reviewed by David Capwell and Caleb Rackliffe for CASSANDRA-18302

maedhroz committed:
fixing remaining (mostly compilation) issues after initial rebase of cep-15-accord on cep-21-tcm

dcapwell committed:
Operations.migrateReadRequiredOperations fails due to concurrent access when TransactionStatement is prepared

patch by David Capwell; reviewed by Ariel Weisberg, Caleb Rackliffe for CASSANDRA-18337

dcapwell committed:
CEP-15 (Accord) Expected reply message with verb ACCORD_INFORM_OF_TXNID_RSP but got ACCORD_SIMPLE_RSP

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18375

dcapwell committed:
CEP-15 (Accord) Original and recover coordinators may hit a race condition with PreApply where reads and writes are interleaved, causing one of the coordinators to see the writes from the other

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18422

belliottsmith committed:
(Accord only) Permit nodes to join a cluster without the full transaction history

patch by Benedict; reviewed by Blake Eggleston for CASSANDRA-18523

dcapwell committed:
CEP-15: (C*) Accord message processing should avoid being passed on to a Stage and run directly in the messageing handler

patch by David Capwell; reviewed by Ariel Weisberg, Benedict Elliott Smith for CASSANDRA-18364

aweisberg committed:
Accord/non-Accord interoperability

Patch by Ariel Weisberg and Blake Eggleston; Reviewed by Blake Eggleston for CASSANDRA-18129

Co-authored-by: Blake Eggleston <beggleston@apple.com>

dcapwell committed:
CEP-15: (C*) Enhance in-memory FileSystem to work with mmap and support tests to add custom logic

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18485

dcapwell committed:
CEP-15: (C*) Add notion of CommandsForRanges and make this durable in C*

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18519

Commit:
minor cleanup in Verb and BTreeSet

Commit:
removing unnecessary calls to ServerTestUtils.daemonInitialization() in a handful of tests

Commit:
make sure workspace.xml specifies a storagedir

iamaleksey committed:
CEP-15: Extend Accord MessageType with a side effect flag

patch by Aleksey Yeschenko; reviewed by Benedic Elliott Smith for
CASSANDRA-18561

iamaleksey committed:
CEP-15: Convert AccordStateCache cache from write-through to write-back

patch by Aleksey Yeschenko; reviewed by Blake Eggleston for
CASSANDRA-18563

bdeggleston committed:
CEP-15: Accord Bootstrap Integration

Patch by Blake Eggleston and Benedict Elliott Smith; Reviewed by David
Capwell for CASSANDRA-17101

CEP-15: Accord TCM integration

Patch by Blake Eggleston; Reviewed by David Capwell for CASSANDRA-18444

iamaleksey committed:
CEP-15: Minimize transaction state kept in system tables

patch by Aleksey Yeschenko; reviewed by Ariel Weisberg for CASSANDRA-18573

jonmeredith committed:
Several simulator fixes not yet merged to cep-15-accord

s c4bf1533b4 Add more JVM arg checks to the paxos simulation runner
s 1b650a0636 break circular dependency between FileSystemOwnershipCheck and CassandraRelevantProperties
s e0111748dd fix verify with-rng and interrupts
s f8475f2a18 make SimulatedFailureDetector deterministic
s ee77734bbe Deterministic transformation
s 890f23696a Add mechanism forsimulator to intercept password salting
s 5205b99874 Specify Paxos V2 because it is the only linearizable option

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19008

belliottsmith committed:
Support state eviction (i.e. permit the state machine to erase transactions that are known to be applied across the cluster)

patch by Benedict Elliott Smith; reviewed by Ariel Weisberg, Aleksey Yeschenko, and David Capwell for CASSANDRA-18883

Co-authored-by: Benedict Elliott Smith <benedict@apache.org>
Co-authored-by: Ariel Weisberg <aweisberg@apple.com>
Co-authored-by: Aleksey Yeschenko <aleksey@apache.org>
Co-authored-by: David Capwell <dcapwell@gmail.com>

dcapwell committed:
CEP-15 (C*) When a host replacement happens don't loose the peer mapping right away (#3575)


patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-18764

dcapwell committed:
Fix Accord compaction purger tombstone logic

Accord compaction purgers see random slices of Accord state during compaction (based on randomly selected compaction inputs).

For at least the `durability` column in the `commands` table the tombstone being created when truncating was deleting the latest value since we can get enough information to truncate without actuall yhaving the latest `durability` value.

To fix we can wait to emit a tombstone until we are erasing the entire command row when truncating or truncating with outcome and meanwhile we can drop the extra columns that are no longer needed instead of using a tombstone. We don't need to emit cell tombstones we can drop them from the purger when processing each row.

patch by Ariel Weisberg; reviewed by David Capwell for CASSANDRA-18795

jacek-lewandowski committed:
CASSANDRA-18774: Fix pre-commit hook

aweisberg committed:
Allow exceptions to be propagated remotely

https://github.com/apache/cassandra-accord/pull/56

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-18779

dcapwell committed:
CEP-15 (C*): when loading commands that have empty waiting_on, make sure not to loose the partial deps (#3590)

patch by David Capwell; reviewed by Aleksey Yeschenko for CASSANDRA-18783

iamaleksey committed:
Fix CompactionAccordIteratorsTest

bdeggleston committed:
Bootstrap/TCM integration test fixes

dcapwell committed:
CEP-15: (C*) NPE on restart as epoch isnt known but verb handler doesnt delay mapping to Node.Id

dcapwell committed:
CEP-15: (C*) accord.messages.BeginRecovery.RecoverNack#supersededBy is nullable but C* serializer doesn't expect null

Commit:
- Avoid persisting fragments that do not require completion as Updates in TxnWrite, as they can simply be pulled from PartialTxn when needed in Write#apply()
- Avoid serializing full TxnData instances to Accord state tables

patch by Caleb Rackliffe; reviewed by David Capwell, Benedict Elliot Smith, and Ariel Weisberg for CASSANDRA-18355

maedhroz committed:
ninja-fix: checkstyle fixes in VirtualTableTest, HistoryValidatorTest, SeedDefiner, RunStartDefiner, and Config

iamaleksey committed:
Fix Propagate serializer

maedhroz committed:
ninja-fix: minor post-TCM-rebase cleanup

maedhroz committed:
ninja-fix: exclude accord_demo.txt from rat checks

maedhroz committed:
Fix AccordJournalTest and ConfigCompatibilityTest, and revert the sendMessageWithFailureCB() callback to printing just the repair failure reason in its error message

dcapwell committed:
Get simulator working (again)

Co-authored-by: Ariel Weisberg <aweisberg@apple.com>
Co-authored-by: Benedict Elliott Smith <benedict@apache.org>

jacek-lewandowski committed:
CEP-15: Add Accord configuration stub

Patch by Jacek Lewandowski; reviewed by David Capwell for CASSANDRA-18221

belliottsmith committed:
Improve validation and address various discovered faults

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-19045

jacek-lewandowski committed:
CEP-15: Accord metrics

Patch by Jacek Lewandowski, reviewed by Caleb Rackliffe, David Capwell and Henrik Ingo for CASSANDRA-18580

dcapwell committed:
CEP-15: (C*) Implement TopologySorter to prioritise hosts based on DynamicSnitch and/or topology layout

patch by Blake Eggleston, David Capwell; reviewed by Blake Eggleston for CASSANDRA-18929

ifesdjeen committed:
Use pinned Harry version

bdeggleston committed:
Reduce command deps

Patch by Blake Eggleston; Reviewed by Benedict Elliott Smith for CASSANDRA-18784

aweisberg committed:
Ninja fix SlowMessageFuzzTest and InvertedIndexSearcherTest

Patch by Ariel Weisberg; Reviewed by David Capwell for CASANDRA-19023

aweisberg committed:
Fix Paxos V2 prepare response serialization

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19023

iamaleksey committed:
Pre-requisite changes for CASSANDRA-18888

patch by Aleksey Yeschenko; reviewed by Ariel Weisberg for
CASSANDRA-18888

iamaleksey committed:
Quick fix for AccordCommandStoreTest#commandLoadSave()

dcapwell committed:
Ninja for CASSANDRA-19045: make sure to use https rather than git@ for submodule so its portable

dcapwell committed:
Ninja for CASSANDRA-19045: use the latest sha from trunk rather than an old one from 10 months ago

aweisberg committed:
Fix additional live migration/interop merge issues

s fc83325fa9 Fix AccordObjectSizes empty sizes
s 44bce6a08c Fix error handling when there are no column families to repair
s f1459540e5 Fix broken nowInSec deserialization
s 91c1befd1c Fix Mutation serializedSize
s b11467c200 Using simulated clock instead of raw nanoTime
s 6412b35924 Accord repair needs to use sentinel tokens
s d19c01ff01 Enable Accord repair with HappyPathFuzzTest
s 2da87f91b8 Implement abort in AccordRepairJob
s e12877c4fa move barrier to accord spec

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19023

aweisberg committed:
Fix AccordCommandsPurger universal durability check

aweisberg committed:
Fix AccordJournal.Type interop entries

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19024

bdeggleston committed:
Fix empty row check in command/tfk mutation methods

bdeggleston committed:
Fix CompactionAccordIteratorsTest

aweisberg committed:
Fix Mutation serializer

aweisberg committed:
Fix 5.1 rebase bugs

aweisberg committed:
Add back enforceInvariants=false to shouldCleanup in AccordCommandsPurger

aweisberg committed:
Fix StorageAttachedIndexDDLTest failing due to background Accord compactions

aweisberg committed:
Instance.receiveMessage should use sync

We mistakenly set this to async thinking async delivery was needed, but whether it is async or not is based on the boolean paramter passed to it.

Right now SchemaTest relies on it running sync, deserializaing in the other class loader and generating an error, and then handling that error from inside C*. This isn't how it actually works on failed deserialize (it's just a timeout), but the test relies on it so for now not changing it.

aweisberg committed:
Fix GuardrailTablesTest

aweisberg committed:
Don't use ImmediateExectuor in JVM dtests to process messages

For now just throwing it on MISC stage which is an unbounded thread pool that shouldn't deadlock

aweisberg committed:
Fix TombstoneCountWarningTest

bdeggleston committed:
Schema based accord fast path configuration

Patch by Blake Eggleston; Reviewed by David Capwell and Alex Petrov for CASSANDRA-19009

dcapwell committed:
(Accord): Bug fixes from CASSANDRA-18675 to better support adding keyspaces

patch by David Capwell; reviewed by Benedict Elliott Smith, Blake Eggleston for CASSANDRA-18804

dcapwell committed:
(Accord) NPE while trying to serialize FoundKnownMap as value is null half the time but unexpected while serializing

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-19253

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1187,6 +1187,47 @@ public static void mutateMV(ByteBuffer dataKey, Collection<Mutation> mutations,
+        return result;
+    }
+
+    private static boolean writesThroughAccord(List<? extends IMutation> mutations)
```
This isn't range based so the entire table is opted into Accord so it's not possible to limit the blast radius to a single key.

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/schema/TableAttributes.java:
```diff
@@ -165,6 +167,14 @@ private TableParams build(TableParams.Builder builder)
         }
 
+        // TODO (now): fail if attempting to change a system table
+        // TODO (now): set migration to complete if this is a new table
```
A pretty valuable/important to do, do you want to do it in this PR?

bdeggleston replied:
handled both in create/alter table statements

aweisberg commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamReceiver.java:
```diff
@@ -246,7 +246,7 @@ public void finished()
         IAccordService accordService = AccordService.instance();
         if (session.streamOperation().requiresBarrierTransaction()
-            && accordService.isAccordManagedTable(cfs.getTableId())
+            && cfs.metadata().isAccordEnabled()
```
I just realized this probably breaks if we try to run the barrier on a range that is only partially Accord managed. I'll bet Accord generates an error. This should check if the ranges actually intersect with a migrating/migrated range and only run the barrier to limit the blast radius of using Accord.

It will also race with user modifications to the migrating ranges so uh yuck, but should also probably retry a few times. We already have a retrying barrier method in `AccordService` that should probably be modified to handle this error and take a strategy of either bubble up the error or retry with just the intersecting ranges managed by Accord.

aweisberg replied:
Are you going to do anything for this as part of your incremental repair changes?

bdeggleston replied:
to clarify, have you opened a JIRA that would address this comment? Excluding ranges that aren't migrating wouldn't be too hard here, but am fine with it being deferred to another ticket

aweisberg replied:
I opened a JIRA. It's actually more complicated because of the need for retries when racing with topology changes.

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/schema/AlterTableStatement.java:
```diff
@@ -553,6 +554,33 @@ public void validate(ClientState state)
+                return next;
+
+            if (wasMigrating)
+                throw ire(format("Cannot change transactional mode from %s to %s for %s.%s before transactional migration has completed",
```
They should be able to change between the different interop options if migrating to Accord?

I think the only restriction we want to add here (and via JMX) is that they can be migrating, but there can be no migrating ranges because it's not safe to flip the direction of a range that starts but doesn't compelte migration.

bdeggleston replied:
I think we'd need a migration to occur between the different interop modes as well? Like if we moved from `full` to `migration`, I think we'd want to make sure data written with `full` had been repaired so it would be visible to non-serial reads?

aweisberg replied:
You are right some of them have should require a migration, but you should be able to change interop modes.

From Accord's perspective the only mode that matters is `full` (ignores supplied consistency level, does async commit), and not `full`.

For C* the modes you should be able to switch between unsafe <-> mixed <-> migration. Really `migration` is the only mode that you can safely use on anything. The rest make the assumption the application is not going to mix Accord/non-Accord reads/writes on the same data.

aweisberg replied:
You are right some of them have should require a migration, but you should be able to change interop modes.

From Accord's perspective the only mode that matters is full (ignores supplied consistency level, does async commit), and not full.

For C* the modes you should be able to switch between unsafe <-> mixed <-> migration. Really migration is the only mode that you can safely use on anything. The rest make the assumption the application is not going to mix Accord/non-Accord reads/writes on the same data.

bdeggleston replied:
This doesn't prevent you from changing interop modes, but it does prevent you from changing interop mode while there's an unfinished migration. 

Since each of the migration modes provide different consistency guarantees, I think a migration between at least some of them is needed to prevent invalidating programmer assumptions in between them. For instance, you can do normal QUORUM reads during the “migration” (mixed_reads proposed) interop mode. We’ll want to at least run a barrier as a migration so that successfully completed writes don’t become intermittently unreadable since the writes at `full` may have only been persisted on a single node.

When upgrading from an interop mode allowing unsafe writes, don’t think it would be a bad idea to keep the table in a symbolic migrating state for operator visibility and alerting purposes, since the safety of the safer interop mode won’t be available until data is repaired.

As an aside I’d thought the “mixed” interop mode did something like a quorum read for accord operations to prevent non-serial writes from causing multiple outcomes for accord operations, but it looks like I misremembered that (or am not looking hard enough). Assuming I’m not missing something, we should definitely rename mixed to unsafe_writes since non-serial writes can cause multiple outcomes, it’s just a little less dangerous since BRR is routed through accord.

aweisberg replied:
OK, I can sort of see that if you are running an unsafe mode in Cassandra it would require a restart of the migration to actually migrate with the level of safety/correctness of the safe mode (`migration`).

I think the pain point is if you set an unsafe mode and want to switch to a safe mode and restart the migration then you can't. You have to complete the unsafe one which requires a repair to complete and you are exposed to data loss the entire time.

Force migration change still works so I think that is good enough for now.

Accord always respects the consistency levels supplied except in `TransactionMode.full` where everything runs through Accord. It's Cassandra that has all the unsafe ways to operate to try and preserve performance assuming that you won't overlap access in unsafe ways.

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/schema/CreateTableStatement.java:
```diff
@@ -125,6 +126,10 @@ public Keyspaces apply(ClusterMetadata metadata)
         if (!table.params.compression.isEnabled())
             Guardrails.uncompressedTablesEnabled.ensureEnabled(state);
 
+        if (table.params.transactionalMode.accordIsEnabled && !DatabaseDescriptor.getAccordTransactionsEnabled())
```
Do we have a way for to default tables into Accord being enabled on creation?

bdeggleston replied:
there's not, do you think we need a config option?

aweisberg replied:
I think it's pretty important yeah since otherwise you have to be very careful when creating schema and if you aren't you have to run the migration process. If the migration process were turn key it would be much less of an issue.

aweisberg commented on src/java/org/apache/cassandra/service/accord/AccordService.java:
```diff
@@ -706,4 +669,27 @@ public Pair<Int2ObjectHashMap<RedundantBefore>, DurableBefore> getRedundantBefor
         return Pair.create(redundantBefores, durableBefore.get());
     }
+
+    public static void ensureTableIsEnabledForAccord(TableMetadata metadata)
```
These all assert the table is enabled for Accord they don't ensure it which at least to me implies that they actually enable it.

aweisberg commented on src/java/org/apache/cassandra/service/consensus/TransactionalMode.java:
```diff
@@ -0,0 +1,136 @@
+ * // We could do it by range instead in the migration version, but then we need to know when all in flight writes
+ * // are done before marking a range as migrated. Would waiting out the timeout be enough (timeout bugs!)?
+ */
+public enum TransactionalMode
```
I think this name is problematic.

The current set of settings are kind of idiosynchratic to Accord where mixed is actually still unsafe. Maybe it just shouldn't exist, but I wanted to have it in there so you aren't forced to run all writes through Accord in a high write throughput use cases.

Another concern is that we also support LWT which are transactions even if they are not very good ones.

Interoperability needs to be broken out from enabled/disabled as mentioned in the `AlterTable` comment. It also seems like we want a default so you only have to specify it in unusual situations where you want to trade safety for performance or vice versa from the default.

Defaults are going to be important when you don't own the schema and you want to roll it out widely to existing applications that will be creating and destroying tables outside of our control.

bdeggleston replied:
Do you mean the name TransactionalMode? How is it problematic? 

If you mean the individual setting names, I'd agree they could be improved a bit. `mixed` could be called `mixed_unsafe`. I'm not sure what `migration` should be called.

It's not clear how reads are handled for each of these in either case.

Do you have a suggestion for how interop should be broken out? Do you mean as a 3rd schema option?

It sounds like we will want configurable defaults for on/off and/or interop mode that are applied in CreateTableStatement

aweisberg replied:
The name `TransactionalMode` is problematic because this is really hyper focused on interoperability with Accord and we have `LWT` so the word `Transactional` makes it seem like this is how `LWT`s are configured and it's not just that. It's really `AccordInteroperabilityMode`.

I was thinking it might be cleaner to just have an on/off for interoperability, but I have warmed up to the current approach. I just think maybe allowing changing between different interoperability modes should be allowed. More in my comment on alter table.

bdeggleston replied:
> It's really AccordInteroperabilityMode.

Its main function is turning accord on or off for a table. Interop modes are included here to avoid having a separate config setting that only makes sense if this one is on

Regarding the naming of the name and/or it's settings, do you have any suggestions?

bdeggleston replied:
for mode names, how about: `off, unsafe, unsafe_writes, mixed_reads, full`? Tthat seem like it more accurately describes the tradeoffs being made w/ each mode

bdeggleston replied:
Regarding the name TransactionalMode, I see the names Transactional and Accord as having different functions. One refers to the C* feature and and the other refers to the consensus protocol, respectively. Since configuration is user facing, I don’t think naming it accord is appropriate (and think AccordSpec should be renamed to TransactionalSpec). This is similar to how we have a user feature call lwts, and they are implemented w/ paxos.

aweisberg replied:
OK, that makes sense.

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1187,6 +1187,47 @@ public static void mutateMV(ByteBuffer dataKey, Collection<Mutation> mutations,
+    private static ConsistencyLevel consistencyLevelForCommit(Collection<? extends IMutation> mutations, ConsistencyLevel consistencyLevel)
+    {
+        ConsistencyLevel result = null;
+        for (IMutation mutation : mutations)
```
`commitCLForStrategy` returns `null` instead of the supplied consistency level if a table is configured to fully use Accord. This would means an earlier iterated table that requires it's consistency level to be honored wouldn't get that.

Additionally it's possible to break an application by enabling migration for one table and not another and we don't support atomically updating multiple tables at once so users can't even do it correctly anyways.

Really what needs to happen is break apart the mutations into ones that need to run through Accord and ones that need to run outside of Accord. This isn't a new deficiency it's just something needs to be added to interop.

bdeggleston replied:
We'll have to give this some thought, since that will mean splitting up mutations that write to multiple tables, as well as making write coordination concurrent so that we coordinate normal and accord writes simultaneously

bdeggleston replied:
I think that might be enough of an undertaking to warrant punting it to a follow on ticket. I also wonder if we could get away with just having this restriction for interop.... basically forbidding multi-table / multi-key batches when using accord, or migrating to/from it. It is something we discourage

aweisberg replied:
I recall from discussions about some use cases and reviewing their queries that they definitely use unlogged batches and the use case is to split data over multiple partitions so they don't run into the large partition problem.

I would need to resurrect the full query log I have to confirm.

aweisberg replied:
I confirmed that some use cases definitely do batches to multiple tables and partitions from the full query log I took a while back.

bdeggleston replied:
deferred to https://issues.apache.org/jira/browse/CASSANDRA-19431

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1187,6 +1187,47 @@ public static void mutateMV(ByteBuffer dataKey, Collection<Mutation> mutations,
+        if (!accordWrite && !normalWrite)
+            throw new IllegalStateException("Mutation is neither a normal write nor an accord write");
+        if (accordWrite && normalWrite)
+            throw new InvalidRequestException("Cannot mix writes to accord and normal tables");
```
As mentioned above I don't think we can safely avoid supporting that.

bdeggleston replied:
deferred to https://issues.apache.org/jira/browse/CASSANDRA-19431

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1220,12 +1261,12 @@ public static void mutateWithTriggers(List<? extends IMutation> mutations,
-        NonSerialWriteStrategy nonSerialWriteStrategy = DatabaseDescriptor.getNonSerialWriteStrategy();
-        if (nonSerialWriteStrategy.writesThroughAccord && !SchemaConstants.getSystemKeyspaces().contains(keyspaceName))
-            mutateWithAccord(augmented != null ? augmented : mutations, consistencyLevel, queryStartNanoTime, nonSerialWriteStrategy);
+        if (writesThroughAccord(mutations) && !SchemaConstants.getSystemKeyspaces().contains(keyspaceName))
```
Maybe the `SchemaConstants` check should now be an assertion. This check could also be in create and alter table.

aweisberg replied:
This check for the system keyspaces is gone now. Should still be kept around as an assertion in case we have test code or otherwise that ends up here?

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1252,11 +1293,11 @@ private static void mutateWithAccord(Collection<? extends IMutation> iMutations,
-        AccordUpdate update = new TxnUpdate(fragments, TxnCondition.none(), clForCommit);
+        ConsistencyLevel clForCommit = consistencyLevelForCommit(mutations, consistencyLevel);
+        TxnUpdate update = new TxnUpdate(fragments, TxnCondition.none(), clForCommit);
+        logger.info("Mutating with Accord clForCommit " + clForCommit);
```
Remove before commit

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -1651,25 +1649,17 @@ public List<Integer> finishConsensusMigration(@Nonnull String keyspace,
     {
         ClusterMetadata cm = ClusterMetadata.current();
-        ConsensusMigrationState snapshot = cm.consensusMigrationState;
+        MigrationState snapshot = cm.consensusMigrationState;
```
`MigrationState` is not a great name just because in the lifecycle of Cassandra we might have more migration type things.

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -1651,25 +1649,17 @@ public List<Integer> finishConsensusMigration(@Nonnull String keyspace,
+        // TODO wanted something human and machine readable, but didn't expend a lot of thought
+        // on what human readable conventions should be, also couldn't get snakeyaml
+        // to output YAML containing only primitives without going through the goofy mapping process
+        // it adds tags and references that clutter the output up badly
```
Does this comment need to come back?

aweisberg commented on src/java/org/apache/cassandra/service/accord/txn/TxnQuery.java:
```diff
@@ -154,10 +152,6 @@ public Result compute(TxnId txnId, Timestamp executeAt, Seekables<?, ?> keys, @N
         Epoch epoch = Epoch.create(executeAt.epoch());
         if (transactionIsInMigratingOrMigratedRange(epoch, keys))
         {
-            // Fail fast because we can't be sure where this request should really run or what was intended
```
This check could stay since we still know the tables from the keys and check their metadata. Feels borderline since it means you were able to alter the table unsafely.

bdeggleston replied:
I'd lean towards leaving it out, and I'm not sure what it would look like if we left it. The original check seemed to be more an artifact of LWTs being converted and interop/migration being implemented independently of each other. With the schema config, there isn't really an LWT strategy, your table is either using accord or not, and that, combined with migration state,  informs how LWTs are handled.

aweisberg replied:
The check if whether the key is in a migrating or migrated range from Accord to Paxos, so if the strategy for the table is `TransactionalMode.full` then the key shouldn't be in a migrating or migrated range *from* Accord and still be attempting to execute on Accord.

So is the check useful? It's a bit mid.

This is also another place where we are using the `ConsensusMigrationState` from the current cluster metadata and not the epoch the query executes at! I have a patch set that fixes this in another place I need to update.

bdeggleston replied:
> The check if whether the key is in a migrating or migrated range from Accord to Paxos, so if the strategy for the table is TransactionalMode.full then the key shouldn't be in a migrating or migrated range from Accord and still be attempting to execute on Accord.

If we do migrations between interop modes then this check would be overly defensive

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/ConsensusMigrationTarget.java:
```diff
@@ -0,0 +1,62 @@
+{
+    paxos(0),
+    accord(1),
+    reset(2);  // FIXME: remove
```
This still seems like a pretty valid escape hatch when you want to get out of migration and just pick a specific system.

bdeggleston replied:
in that case you'd just manually set the tranaction migration mode to `non` for the table

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/ConsensusRequestRouter.java:
```diff
@@ -76,43 +74,56 @@ public static void resetInstance()
-        ClusterMetadata cm = ClusterMetadata.current();
 
-        TableMigrationState tms = cm.consensusMigrationState.tableStates.get(cfs.getTableId());
+        if (tmd.params.transactionalMigrationFrom == TransactionalMigrationFromMode.none)
```
Took me a while to figure out that `transactionalMigrationFrom`  `none` means no migration in progress anymore. Maybe call it `none_in_progress`?

I know it's obvious in retrospect.

bdeggleston replied:
If you don't mind, I'd like to keep the mode name the same, for DDL compactness. I did add an `isMigrating` method to `TransactionalMigrationFromMode` and replaced checks against `none` with it to aid readability. WDYT?

aweisberg replied:
WFM

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/ConsensusTableMigrationState.java:
```diff
@@ -1,907 +0,0 @@
-/*
```
Would be better to use `git mv` to preserve the history.

Were there changes to the other classes pulled of `ConsensusTableMigrationState` that I should check for, or are they just being extracted?

bdeggleston replied:
yeah it was split into multiple files though. I don't _think_ there were any major changes, but I'd give it a once over to be sure

aweisberg commented on src/java/org/apache/cassandra/service/reads/repair/BlockingReadRepair.java:
```diff
@@ -193,9 +192,8 @@ public void awaitWrites()
-        if (coordinator.isEventuallyConsistent() && (DatabaseDescriptor.getLWTStrategy() == LWTStrategy.accord
-                                                     || nonSerialWriteStrategy.blockingReadRepairThroughAccord))
+        TransactionalMode transactionalMode = command.metadata().params.transactionalMode;
+        if (coordinator.isEventuallyConsistent() && transactionalMode.blockingReadRepairThroughAccord)
```
This also should do read repair separately for Accord/non-Accord ranges 😭

bdeggleston replied:
assuming this will be done as part of https://issues.apache.org/jira/browse/CASSANDRA-19431

aweisberg commented on src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java:
```diff
@@ -249,22 +253,56 @@ private static Map<ReplicationParams, Set<KeyspaceMetadata>> groupByReplication(
+                    break;
+            }
+
+            migrationState = migrationState.withRangesMigrating(startedIds, ranges, true);
```
When changing directions in a migration we don't want to do `auto` if all ranges weren't already migrated since that repeats work already done. I think we want to invert the migration using `tableMigrationState.withMigrationTarget`.

aweisberg replied:
Also this is a correctness issue because if you don't precisely flip the ranges then you are basically migrating ranges to the new "from" that never actually finished migration.

In fact I think that the current approach for flipping directions is unsafe if we allow it to occur when any range hasn't been fully migrated because the range now consists of a mix of migrated keys and ranges waiting to be migrated and the ranges waiting to be migrated can't be forcibly swapped to the new system because then we will reverse the kind of key migration we are doing without doing a repair.

aweisberg replied:
Actually I think this is safe/fine since all keys already migrated will be key migrated before being used. In fact you could reverse several times and it should be fine? You will migrate on one system before using the other and you are always going to see the latest value for a key because either you picked the correct system, or you did the key migration first.

aweisberg replied:
Although one random thought is that it's kind of important that the timestamps published by Paxos and Accord have the necessary barriers. IOW Accord should never write a value with a timestamp lower than what Paxos has used.

aweisberg commented on src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java:
```diff
@@ -249,22 +253,56 @@ private static Map<ReplicationParams, Set<KeyspaceMetadata>> groupByReplication(
+            migrationState = migrationState.withMigrationsCompletedFor(droppedIds);
+
+        Set<TableId> completedIds = altered.stream().flatMap(diff -> diff.tables.altered.stream())
+                                           .filter(alt -> alt.before.params.transactionalMigrationFrom != TransactionalMigrationFromMode.none
```
Is this the case where they manually altered `transactionalMigrationFrom` and forced the completion of the migration by just saying "well it's not migrating?"

We should log the ranges that were migrating so when they do this unsafely there is some indication of what might have been impacted and that this unsafe action was performed.

Borderline it shouldn't be allowed unless an explicit option is set on the C* process. We will have real headaches if people change this and then have problems and we lose the logs and don't know what happened.

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/schema/AlterTableStatement.java:
```diff
@@ -553,6 +554,33 @@ public void validate(ClientState state)
+                                 keyspaceName, tableName));
+
+            // set table to migrating
+            TransactionalMigrationFromMode migrateFrom = TransactionalMigrationFromMode.fromMode(prev.transactionalMode, next.transactionalMode);
```
It's very confusing here that users can supply this, but then we replace it.

bdeggleston replied:
we only set it if they haven't supplied it, otherwise we return above

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/schema/AlterTableStatement.java:
```diff
@@ -553,6 +554,33 @@ public void validate(ClientState state)
+        {
+            boolean modeChange = prev.transactionalMode != next.transactionalMode;
+            boolean wasMigrating = prev.transactionalMigrationFrom != TransactionalMigrationFromMode.none;
+            boolean forceMigrationChange = prev.transactionalMigrationFrom != next.transactionalMigrationFrom;
```
Is this meant to be the unsafe way to make a migration complete?

bdeggleston replied:
yes, or otherwise manually intervene

aweisberg replied:
I requested this elsewhere earlier, but can we log any time a forcible migration change happens since it's usually going to be some unsafe combination?

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TableMigrationState.java:
```diff
@@ -0,0 +1,345 @@
+    public final TableId tableId;
+
+    @Nonnull
+    public final ConsensusMigrationTarget targetProtocol;
```
We definitely want to remove this, and get it from the `TableMetadata`. If we store it twice we are not going to keep it in sync.

bdeggleston replied:
I think I'd prefer to keep this as is. Since epochs can change during the processing of a query, I think it will be less work to keep this in sync than it will be to ensure you're looking at the migration state and table metadata from the same epoch. I added a sanity check to the ClusterMetadata build method to check that they're in sync

aweisberg commented on src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java:
```diff
@@ -249,22 +253,56 @@ private static Map<ReplicationParams, Set<KeyspaceMetadata>> groupByReplication(
+                                         .collect(toImmutableSet());
+
+        if (!droppedIds.isEmpty())
+            migrationState = migrationState.withMigrationsCompletedFor(droppedIds);
```
This keeps state around for dropped tables. We should remove them entirely the way it used to?

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/MigrationState.java:
```diff
@@ -0,0 +1,218 @@
+        {
+            tableState = new TableMigrationState(metadata.keyspace, metadata.name, metadata.id, target, ImmutableSet.of(), ImmutableMap.of(Epoch.EMPTY, ranges));
+        }
+        next.put(metadata.id, tableState);
```
Does this actually work? I thought `ImmutableMap.Builder` throws an exception on duplicate keys when you call `build()`.

There is apparently a `buildKeepingLast` method which should work?

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/MigrationState.java:
```diff
@@ -0,0 +1,218 @@
+        return new MigrationState(lastModified, updated.build());
+    }
+
+    public MigrationState withRangesMigrating(TableMetadata metadata, List<Range<Token>> ranges, boolean overwrite)
```
Unused?

aweisberg commented on src/java/org/apache/cassandra/tcm/transformations/MaybeFinishConsensusMigrationForTableAndRange.java:
```diff
@@ -122,9 +145,15 @@ public Result execute(@Nonnull ClusterMetadata metadata)
+        next = next.with(migrationState);
+
+        // reset the migration value on the table if the migration has completed
+        if (migrationState.tableStates.containsKey(tbm.id))
```
Why is just having a `migrationState` enough to indicate that migration has completed entirely?

aweisberg reviewed (changes requested):
Finished reviewing up to the tests. Found several issues, some were pre-existing problems with interop not being range based, others are unrelated.

It looks like it's mostly there UI wise in terms of getting it into the table schema. and having all the necessary functionality.

I personally had a really hard time figuring out the relationship between the migration from stuff and the transactional mode, but maybe that is just me.

aweisberg commented on test/distributed/org/apache/cassandra/distributed/test/accord/AccordCQLTest.java:
```diff
@@ -2422,13 +2414,13 @@ private void testSelectUDTReference(String tableDDL) throws Exception
     public void testSelectMultiCellUDTFieldReference() throws Exception
     {
-        testSelectUDTFieldReference("CREATE TABLE " + qualifiedTableName + " (k int PRIMARY KEY, customer person)");
+        testSelectUDTFieldReference("CREATE TABLE " + qualifiedTableName + " (k int PRIMARY KEY, customer person) WITH transactional_mode='" + transactionalMode + "'");
```
This test is a strong argument in favor of allowing a default to be set. Maybe a `keyspace` level thing? Or `Config`.

aweisberg commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -1112,8 +1111,13 @@ public enum PaxosOnLinearizabilityViolation
+     * explicit, no ranges are marked as migrating, and the user needs to explicitly mark ranges as migrating to
+     * the target transactional mode via nodetool.
+     */
+    public volatile TransactionalRangeMigration transactional_range_migration = TransactionalRangeMigration.auto;
```
Should this go in `AccordSpec`?

aweisberg commented on test/unit/org/apache/cassandra/schema/TransactionalSchemaTest.java:
```diff
@@ -0,0 +1,104 @@
+
+import static java.lang.String.format;
+
+public class TransactionalSchemaTest
```
The test name makes it sound like it is testing the ability to transactionally modify schema when it's really testing test table level transaction config.

The test doesn't validate that it connects to the rest of the request routing stuff correctly. Maybe `MigrationTest` should be parameterized so that it can use `ALTER TABLE` where possible so we cover both JMX and alter table?

aweisberg commented on test/unit/org/apache/cassandra/schema/TransactionalSchemaTest.java:
```diff
@@ -0,0 +1,104 @@
+        // explicitly setting the migration mode should work
+        process("ALTER TABLE ks.%s WITH transactional_mode='%s' AND transactional_migration_from='%s'",
+                table, TransactionalMode.off, TransactionalMigrationFromMode.none);
+        assertTransactionalMode(table, TransactionalMode.off, TransactionalMigrationFromMode.none);
```
A lot of this doesn't really validate that it connects to the rest of the request routing stuff correctly. Maybe `MigrationTest` should be parameterized so that it can use `ALTER TABLE` where possible?

aweisberg commented on test/unit/org/apache/cassandra/schema/TransactionalSchemaTest.java:
```diff
@@ -0,0 +1,104 @@
+
+        try
+        {
+            process("ALTER TABLE ks.%s WITH transactional_mode='%s'", table, TransactionalMode.off);
```
This should work? It's the migration direction reversal request.

ifesdjeen committed:
Quick fix for AccordAddTableTest: make rejections catch up

bdeggleston committed:
fix fast path coordinator test

dcapwell committed:
Ninja: Added check that command.route is not null when SPL is trying to inform, saw this event in CI and need more details to know what happened

bdeggleston committed:
retry on Invalidated

bdeggleston committed:
fix ReducingRangeMapSerializer

bdeggleston committed:
improve accord error logging

bdeggleston committed:
don't make noise when fast path reconfigs are rejected

bdeggleston committed:
apply cfk update to both cfk caches

bdeggleston committed:
don't mark initialized cfk instances as updated

bdeggleston committed:
fix fast path update deadlock

belliottsmith committed:
CASSANDRA-18365: Protocol fixes

iamaleksey committed:
Fix LocalPartitioner duplication in *_for_key Accord system tables

patch by Aleksey Yeschenko; reviewed by Marcus Eriksson for
CASSANDRA-19265

bdeggleston committed:
load both key command histories for operations specifying keyhistory ALL

bdeggleston committed:
fix fast path test epoch wait

belliottsmith committed:
Fix SerializerSupport.reconstruct, and introduce RangeFactory

belliottsmith committed:
Fast single-partition "Ephemeral Reads"

Introduce a special kind of non-durable read that provides only per-key linearizable isolation; i.e. strict-serializable isolation for single partition-key reads.
This read creates only a happens-before edge, by collecting dependencies for execution and ensuring that execution happens strictly after these dependencies
have executed, but at no precise time otherwise. So later writes may be witnessed, and if multiple keys are read they may represents different points in time.

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-19305

belliottsmith committed:
Fix (de)serialization of WaitingOn into cache

belliottsmith committed:
Fix simulator logging and clean shutdown

belliottsmith committed:
dtest Instance shutdown (again)

belliottsmith committed:
Refactor CommandsForKey for efficiency, and to support transitive dependency elision

patch by Benedict; reviewed by Aleksey Yeshchenko for CASSANDRA-19310

aweisberg committed:
Add TODO for TxnQuery using wrong cluster metadata

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+
+import org.apache.cassandra.service.consensus.TransactionalMode;
+
+public enum TransactionalMigrationFromMode
```
Maybe `AccordMigrationFromMode`

and for a comment
```
This tracks the state of a migration either from Paxos -> Accord or Accord -> Paxos. The `TransactionalMode` associated with each transition from a system is how interoperability should be achieved during the migration with various performance/safety tradeoffs.
```

bdeggleston replied:
added comment, will address naming in TransactionalMode thread

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+
+public enum TransactionalMigrationFromMode
+{
+    none(null),
```
```suggestion
    // No migration is in progress. The currently active transaction system could be either Accord or Paxos.
    none(null),
```

aweisberg replied:
Is this comment even technically true? When we finish a migration do we ever actually alter the schema to set this to none or are we basically "always" migrating with the ranges completely migrated and we only ever are in this state when do the very first migration?

aweisberg commented on test/unit/org/apache/cassandra/schema/TransactionalConfigSchemaTest.java:
```diff
@@ -0,0 +1,95 @@
+        process("CREATE TABLE ks.%s (k int primary key, v int)", table);
+        assertTransactionalMode(table, TransactionalMode.off, TransactionalMigrationFromMode.none);
+
+        process("ALTER TABLE ks.%s WITH transactional_mode='%s'", table, TransactionalMode.full);
```
Is this making a table an Accord table with a migration required but without interoperability enabled?

bdeggleston replied:
yes

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+public enum TransactionalMigrationFromMode
+{
+    none(null),
+    off(TransactionalMode.off),
```
```suggestion
    // Unsafe migration from Paxos -> Accord where interoperability is not enabled at all
    off(TransactionalMode.off),
```

bdeggleston replied:
This shouldn't be unsafe. This a transition from accord being off to some accord enabled mode

aweisberg replied:
OK, can you add a comment to that effect?

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+{
+    none(null),
+    off(TransactionalMode.off),
+    unsafe(TransactionalMode.unsafe),
```
```suggestion
    // Migrating back to Paxos from unfinished migration to Accord w/unsafe interop mode
    unsafe(TransactionalMode.unsafe),
```

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+    none(null),
+    off(TransactionalMode.off),
+    unsafe(TransactionalMode.unsafe),
+    mixed(TransactionalMode.mixed),
```
```suggestion
    // Migrating back to Paxos from unfinished migration to Accord w/mixed interop mode
    mixed(TransactionalMode.mixed),
```

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+    off(TransactionalMode.off),
+    unsafe(TransactionalMode.unsafe),
+    mixed(TransactionalMode.mixed),
+    migration(TransactionalMode.migration),
```
```suggestion
    // Migrating back to Paxos from unfinished migration to Accord w/migration interop mode
    migration(TransactionalMode.migration),
```

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TransactionalMigrationFromMode.java:
```diff
@@ -0,0 +1,74 @@
+    unsafe(TransactionalMode.unsafe),
+    mixed(TransactionalMode.mixed),
+    migration(TransactionalMode.migration),
+    full(TransactionalMode.full);
```
```suggestion
    // Migrating from Accord back to Paxos
    full(TransactionalMode.full);
```

aweisberg commented on src/java/org/apache/cassandra/service/consensus/TransactionalMode.java:
```diff
@@ -0,0 +1,136 @@
+ */
+public enum TransactionalMode
+{
+    off(false, false, false, false),
```
```suggestion
    // Running on Paxos V1 or V2 with Accord disabled
    off(false, false, false, false),
```

aweisberg commented on src/java/org/apache/cassandra/service/consensus/TransactionalMode.java:
```diff
@@ -0,0 +1,136 @@
+    unsafe(true, false, false, false),
+
+    /*
+     * Allow mixing of non-SERIAL writes and Accord, but still force BRR through Accord
```
```suggestion
     * Allow mixing of non-SERIAL writes and Accord, but still force BRR through Accord.
     * This mode makes it safe to perform non-SERIAL or SERIAL reads of Accord data, but unsafe
     * to write data that Accord may attempt to read.
```

aweisberg replied:
I got this slightly wrong. It makes it safe to perform SERIAL writes/reads and non-SERIAL reads. The same is true of the "unsafe" mode.

aweisberg commented on src/java/org/apache/cassandra/service/consensus/TransactionalMode.java:
```diff
@@ -0,0 +1,136 @@
+
+    /*
+     * Execute writes through Accord skipping StorageProxy's normal write path, but commit
+     * writes at the provided consistency level so they can be read via non-SERIAL consistency levels.
```
```suggestion
     * writes at the provided consistency level so they can be read via non-SERIAL consistency levels.
     * This mode makes it safe to read/write data that Accord will read/write.
```

aweisberg commented on src/java/org/apache/cassandra/service/consensus/migration/TableMigrationState.java:
```diff
@@ -0,0 +1,360 @@
+        Range<Token> fullRange = new Range<>(partitioner.getMinimumToken(), partitioner.getMinimumToken());
+        List<Range<Token>> allTouched = new ArrayList<>(migratedRanges);
+        allTouched.addAll(migratingRanges);
+        allTouched = Range.deoverlap(allTouched);
```
As far as I can tell this still has the issue that you can reverse migration for partially migrated ranges. For those ranges some keys are on the old system and some are on the new one so it's not a safe operation to reverse direction.

aweisberg replied:
Actually I think this is safe/fine since all keys already migrated will be key migrated before being used. In fact you could reverse several times and it should be fine? You will migrate on one system before using the other and you are always going to see the latest value for a key because either you picked the correct system, or you did the key migration first.

aweisberg commented on src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java:
```diff
@@ -249,22 +249,73 @@ private static Map<ReplicationParams, Set<KeyspaceMetadata>> groupByReplication(
+        if (!completedIds.isEmpty())
+            migrationState = migrationState.withMigrationsCompletedFor(completedIds);
+
+        Map<TableId, TableMetadata> updatedTables = altered.stream()
```
This is unused?

aweisberg commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -1221,116 +1218,9 @@ public enum CorruptedTombstoneStrategy
-     * // are done before marking a range as migrated. Would waiting out the timeout be enough (timeout bugs!)?
-     */
-    public enum NonSerialWriteStrategy
+    public enum TransactionalRangeMigration
```
Maybe this should also get moved into Accord spec. Not really important though.

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/schema/AlterTableStatement.java:
```diff
@@ -553,6 +557,41 @@ public void validate(ClientState state)
+            // user is manually updating migration mode, don't interfere
+            if (forceMigrationChange)
+            {
+                logger.warn("Forcing unsafe migration change from {} to {}", prev.transactionalMigrationFrom, next.transactionalMigrationFrom);
```
Did we also log `transactionalMode` already? Can we include it in this statement?

bdeggleston committed:
CEP-15: (C*) per-table transactional configuration

Patch by Blake Eggleston; Reviewed by Ariel Wesberg for CASSANDRA-19016

bdeggleston commented:
So I took a second look at your requested changes to the TransactionalMigrationFromMode, and didn’t add them since they don’t make sense in the context of accord <-> accord changes we talked about here: https://github.com/apache/cassandra/pull/3032/files#r1446421908
Without the assumption that a migration is always between paxos and accord, the comments don’t say much, so I just left them out. Let me know if you disagree and we can figure something out, I may be overlooking something
